[SPARK-26352][SQL] join reorder should not change the order of output attributes

rednaxelafx opened the pull request:
## What changes were proposed in this pull request?

The optimizer rule `org.apache.spark.sql.catalyst.optimizer.ReorderJoin` performs join reordering on inner joins. This was introduced from SPARK-12032 (https://github.com/apache/spark/pull/10073) in 2015-12.

After it had reordered the joins, though, it didn't check whether or not the output attribute order is still the same as before. Thus, it's possible to have a mismatch between the reordered output attributes order vs the schema that a DataFrame thinks it has.
The same problem exists in the CBO version of join reordering (`CostBasedJoinReorder`) too.

This can be demonstrated with the example:
```scala
spark.sql("create table table_a (x int, y int) using parquet")
spark.sql("create table table_b (i int, j int) using parquet")
spark.sql("create table table_c (a int, b int) using parquet")
val df = spark.sql("""
  with df1 as (select * from table_a cross join table_b)
  select * from df1 join table_c on a = x and b = i
""")
```
here's what the DataFrame thinks:
```
scala> df.printSchema
root
 |-- x: integer (nullable = true)
 |-- y: integer (nullable = true)
 |-- i: integer (nullable = true)
 |-- j: integer (nullable = true)
 |-- a: integer (nullable = true)
 |-- b: integer (nullable = true)
```
here's what the optimized plan thinks, after join reordering:
```
scala> df.queryExecution.optimizedPlan.output.foreach(a => println(s"|-- ${a.name}: ${a.dataType.typeName}"))
|-- x: integer
|-- y: integer
|-- a: integer
|-- b: integer
|-- i: integer
|-- j: integer
```

If we exclude the `ReorderJoin` rule (using Spark 2.4's optimizer rule exclusion feature), it's back to normal:
```
scala> spark.conf.set("spark.sql.optimizer.excludedRules", "org.apache.spark.sql.catalyst.optimizer.ReorderJoin")

scala> val df = spark.sql("with df1 as (select * from table_a cross join table_b) select * from df1 join table_c on a = x and b = i")
df: org.apache.spark.sql.DataFrame = [x: int, y: int ... 4 more fields]

scala> df.queryExecution.optimizedPlan.output.foreach(a => println(s"|-- ${a.name}: ${a.dataType.typeName}"))
|-- x: integer
|-- y: integer
|-- i: integer
|-- j: integer
|-- a: integer
|-- b: integer
```

Note that this output attribute ordering problem leads to data corruption, and can manifest itself in various symptoms:
* Silently corrupting data, if the reordered columns happen to either have matching types or have sufficiently-compatible types (e.g. all fixed length primitive types are considered as "sufficiently compatible" in an `UnsafeRow`), then only the resulting data is going to be wrong but it might not trigger any alarms immediately. Or
* Weird Java-level exceptions like `java.lang.NegativeArraySizeException`, or even SIGSEGVs.

## How was this patch tested?

Added new unit test in `JoinReorderSuite` and new end-to-end test in `JoinSuite`.
Also made `JoinReorderSuite` and `StarJoinReorderSuite` assert more strongly on maintaining output attribute order.

rednaxelafx committed:
SPARK-26352: ReorderJoin should not change the order of columns

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala:
```diff
@@ -895,4 +895,18 @@ class JoinSuite extends QueryTest with SharedSQLContext {
+        with tmp as (select * from tab1 cross join tab2)
+        select * from tmp join tab3 on a = x and b = i
+      """)
+      checkAnswer(df, Row(1, 100, 42, 200, 1, 42))
```
```Scala
scala> spark.sql("with tmp as (select * from tab1 cross join tab2) select * from tmp join tab3 on a = x and b = i").show()
+---+---+---+---+---+---+
|  x|  y|  i|  j|  a|  b|
+---+---+---+---+---+---+
|  1|100| 42|200|  1| 42|
+---+---+---+---+---+---+
```

This still passes without code changes, right?

gatorsmile replied:
Maybe the command lines work but the test case does not work since we turn off `ConvertToLocalRelation` in the test environment?

rednaxelafx replied:
Yeah, before sending this PR I've actually explicitly run both test cases that I added with and without the fix, and verified that without the fix it'd fail the test.

The reason why you "think" it works in your environment is because you used `show`, which after @maropu 's earlier PR will introduce a `Project` at the top, which happens to cure the issue.
If you just do a `collect` you'll see the problem still exists:
```
scala> df.show
+---+---+---+---+---+---+
|  x|  y|  i|  j|  a|  b|
+---+---+---+---+---+---+
|  1|100| 42|200|  1| 42|
+---+---+---+---+---+---+

scala> df.collect
res5: Array[org.apache.spark.sql.Row] = Array([1,100,1,42,42,200])
```

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:
```diff
@@ -99,6 +99,14 @@ object ReorderJoin extends Rule[LogicalPlan] with PredicateHelper {
+      } else {
+        // Reordering the joins have changed the order of the columns.
+        // Inject a projection to make sure we restore to the expected ordering.
+        Project(p.output, reordered)
```
just for curiosity, we only need this for top-level join? I feel it's ok to change the columnar order for intermedia joins.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala:
```diff
@@ -895,4 +895,18 @@ class JoinSuite extends QueryTest with SharedSQLContext {
   }
+
+  test("SPARK-26352: join reordering should not change the order of columns") {
+    withTable("tab1", "tab2", "tab3") {
```
can we reproduce the bug with temp view?

rednaxelafx replied:
That's more prone to be subject to `ConvertToLocalRelation`, i.e. you can't copy-and-paste the test case into a Spark shell and expect the same behavior. Would you still prefer to use temp view instead?

cloud-fan commented:
good catch! LGTM

SparkQA commented:
**[Test build #100057 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100057/testReport)** for PR 23303 at commit [`f0de32f`](https://github.com/apache/spark/commit/f0de32f7d103c7a2522c5278fdb710a8a8d8233f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100057/
Test FAILed.

rednaxelafx committed:
Fix test failures: some test cases themselves are wrong
Also the same output attribute ordering issue is present in the CBO join reordering as well. Fixed it.

SparkQA commented:
**[Test build #100085 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100085/testReport)** for PR 23303 at commit [`0f81933`](https://github.com/apache/spark/commit/0f81933f0ef66ebef5a6bc63444aa5e6f79c6ace).

rednaxelafx commented:
@cloud-fan @gatorsmile @hvanhovell I've updated the PR with a fix to the same problem in CBO join reorder, and also fixed some test cases that have been wrong for a long time (the `expected` plans were in fact wrong).

PTAL, thanks!

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6071/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala:
```diff
@@ -48,8 +48,18 @@ object CostBasedJoinReorder extends Rule[LogicalPlan] with PredicateHelper {
+
+      // Cleanups
       result transformDown {
+        // if a Project was created to keep output attribute order after join reordering, but
```
is this an improvement? Maybe we should do it in another PR, so that this PR can be backported to old branches.

rednaxelafx replied:
This isn't really an improvement. It's here mainly to help pass existing tests -- I'm adding new projections to fix the output attribute order problem, but these extra projections in the middle would make (`expected`) query plans look pretty ugly.

So consider this is two things that can cancel each other out:
1. Add projections to fix output attribute order;
2. If one of these extra projections is in the middle, get rid of it.

The cleanup (2) is only meant to clean extra projections created because of (1). Both (1) and (2) are in this PR so I don't consider this as a performance improvement over existing stuff.

SparkQA commented:
**[Test build #100085 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100085/testReport)** for PR 23303 at commit [`0f81933`](https://github.com/apache/spark/commit/0f81933f0ef66ebef5a6bc63444aa5e6f79c6ace).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100085/
Test FAILed.

gatorsmile commented:
@rednaxelafx Can we separate this to two PRs? One is for the non-CBO reordering. Another is for the CBO reordering?

rednaxelafx commented:
@gatorsmile that was my original intent. I only wanted to fix the non-CBO one in this PR originally, but after reading through the join reorder related test suite I found that most of them have both the non-CBO and the CBO one in the same `Optimize`. The test cases are going to fail one way or the other if we only fix one of them. Sadly we'll have to fix both in the same PR.

The feature fix itself is really small, but granted it's in a very dense piece of code. The majority of this PR is actually about fixing the test cases....a lot of them were wrong to begin with (had incorrect output attribute order in the `expected` plans)

rednaxelafx committed:
Fix StarJoinCostBasedReorderSuite

SparkQA commented:
**[Test build #100110 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100110/testReport)** for PR 23303 at commit [`ae15739`](https://github.com/apache/spark/commit/ae157392c065a84fea05f2e6d323a7c8e3889d8e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6095/
Test PASSed.

rednaxelafx commented:
jenkins retest this please

SparkQA commented:
**[Test build #100117 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100117/testReport)** for PR 23303 at commit [`ae15739`](https://github.com/apache/spark/commit/ae157392c065a84fea05f2e6d323a7c8e3889d8e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6102/
Test PASSed.

SparkQA commented:
**[Test build #100117 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100117/testReport)** for PR 23303 at commit [`ae15739`](https://github.com/apache/spark/commit/ae157392c065a84fea05f2e6d323a7c8e3889d8e).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100117/
Test FAILed.

rednaxelafx commented:
jenkins retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6124/
Test PASSed.

SparkQA commented:
**[Test build #100140 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100140/testReport)** for PR 23303 at commit [`ae15739`](https://github.com/apache/spark/commit/ae157392c065a84fea05f2e6d323a7c8e3889d8e).

rednaxelafx committed:
Move the redudant projection elimination logic from CBO-based join reorder to RemoveRedundantProject

Example: spark.range(10).select('id, 'id, 'id).agg(sum('id))

Before:

== Optimized Logical Plan ==
Aggregate [sum(id#0L) AS sum(id)#9L]
+- Project [id#0L, id#0L, id#0L]
   +- Range (0, 10, step=1, splits=Some(8))

== Physical Plan ==
*(2) HashAggregate(keys=[], functions=[finalmerge_sum(merge sum#13L) AS sum(id#0L)#5L], output=[sum(id)#9L])
+- Exchange SinglePartition
   +- *(1) HashAggregate(keys=[], functions=[partial_sum(id#0L) AS sum#13L], output=[sum#13L])
      +- *(1) Project [id#0L, id#0L, id#0L]
         +- *(1) Range (0, 10, step=1, splits=8)

After:

== Optimized Logical Plan ==
Aggregate [sum(id#0L) AS sum(id)#9L]
 +- Range (0, 10, step=1, splits=Some(8))

== Physical Plan ==
*(2) HashAggregate(keys=[], functions=[finalmerge_sum(merge sum#13L) AS sum(id#0L)#5L], output=[sum(id)#9L])
+- Exchange SinglePartition
   +- *(1) HashAggregate(keys=[], functions=[partial_sum(id#0L) AS sum#13L], output=[sum#13L])
      +- *(1) Range (0, 10, step=1, splits=8)

rednaxelafx commented:
As per offline discussion with @cloud-fan and @hvanhovell , I've updated the PR and moved the logic that eliminates redundant projections from the CBO-based join reorder to `RemoveRedundantProject`.

Could you please take another look? Thanks!
cc @gatorsmile

SparkQA commented:
**[Test build #100162 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100162/testReport)** for PR 23303 at commit [`7ef2fa3`](https://github.com/apache/spark/commit/7ef2fa3a4e49ebac0439c5c7d516cd807478729e).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6141/
Test FAILed.

SparkQA commented:
**[Test build #100162 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100162/testReport)** for PR 23303 at commit [`7ef2fa3`](https://github.com/apache/spark/commit/7ef2fa3a4e49ebac0439c5c7d516cd807478729e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100162/
Test FAILed.

rednaxelafx commented:
Oops, the relaxed transformation in `RemoveRedundantProject` is too wide and it's wrong now. Finding alternatives now.

rednaxelafx committed:
Fix RemoveRedundantProject

SparkQA commented:
**[Test build #100173 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100173/testReport)** for PR 23303 at commit [`385cbd3`](https://github.com/apache/spark/commit/385cbd349b26ef427097d75074693ccff1a19fbb).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6151/
Test PASSed.

SparkQA commented:
**[Test build #100173 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100173/testReport)** for PR 23303 at commit [`385cbd3`](https://github.com/apache/spark/commit/385cbd349b26ef427097d75074693ccff1a19fbb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100173/
Test FAILed.

rednaxelafx committed:
Fix RemoveRedundantProject again: Union needs special handling to make sure all children have matching output attributes

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6154/
Test PASSed.

SparkQA commented:
**[Test build #100176 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100176/testReport)** for PR 23303 at commit [`a7fb7a6`](https://github.com/apache/spark/commit/a7fb7a67e1f4a683f60778f7f66b208ddf77609f).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100176/
Test FAILed.

rednaxelafx commented:
jenkind retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6160/
Test PASSed.

SparkQA commented:
**[Test build #100183 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100183/testReport)** for PR 23303 at commit [`a7fb7a6`](https://github.com/apache/spark/commit/a7fb7a67e1f4a683f60778f7f66b208ddf77609f).

SparkQA commented:
**[Test build #100183 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100183/testReport)** for PR 23303 at commit [`a7fb7a6`](https://github.com/apache/spark/commit/a7fb7a67e1f4a683f60778f7f66b208ddf77609f).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100183/
Test FAILed.

rednaxelafx committed:
Remove the new optimization in RemoveRedundantProject to make this fix less risky

SparkQA commented:
**[Test build #100211 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100211/testReport)** for PR 23303 at commit [`9451b68`](https://github.com/apache/spark/commit/9451b685c6965edcc7f24f677f92fe46831fabe7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100211/
Test PASSed.

gatorsmile commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/StarJoinReorderSuite.scala:
```diff
@@ -395,6 +398,7 @@ class StarJoinReorderSuite extends PlanTest with StatsEstimationTestBase {
         .join(d2.where(nameToAttr("d2_c2") === 2), Inner,
           Some(nameToAttr("f1_fk2") === nameToAttr("d2_pk1")))
         .join(s3, Inner, Some(nameToAttr("f11_fk1") === nameToAttr("s3_pk1")))
+        .select(outputsOf(d1, f11, f1, d2, s3): _*)
```
Could you help update the comment in line 384?

cloud-fan commented:
thanks, merging to master/2.4/2.3!
